File: src/adk_mcp/library.py
import importlib.resources
import json
from typing import List, Dict, Any
import logging
import os

logger = logging.getLogger(__name__)

class ADKLibrary:
    def __init__(self, data_path: str = None):
        self.chunks = []
        self.header_map = {}
        
        if data_path:
             self.data_path = data_path
        else:
            # Använd importlib för att hitta data-filen korrekt inuti paketet
            try:
                # Python 3.9+ stil
                files = importlib.resources.files("adk_mcp.data")
                self.data_path = str(files.joinpath("adk_chunks.jsonl"))
            except Exception:
                # Fallback för äldre system eller om paketet körs direkt från källkod utan installation
                base_dir = os.path.dirname(os.path.abspath(__file__))
                self.data_path = os.path.join(base_dir, "data", "adk_chunks.jsonl")

        self._load_data()

    def _load_data(self):
        try:
            with open(self.data_path, 'r', encoding='utf-8') as f:
                for line in f:
                    try:
                        chunk = json.loads(line)
                        self.chunks.append(chunk)
                        path = chunk.get('header_path', 'Unknown')
                        if path not in self.header_map:
                            self.header_map[path] = []
                        self.header_map[path].append(chunk)
                    except json.JSONDecodeError:
                        logger.warning("Kunde inte avkoda en rad i JSONL-filen.")
        except FileNotFoundError:
            logger.error(f"Filen {self.data_path} hittades inte.")
            pass

    def get_structure(self) -> List[str]:
        """Returnerar en lista av alla rubriker (topics)."""
        return sorted(list(self.header_map.keys()))

    def read_topic(self, topic: str) -> str:
        """Läser allt innehåll för en specifik rubrik."""
        if topic not in self.header_map:
            return f"Error: Topic '{topic}' not found."
        
        return "\n\n".join([c['text'] for c in self.header_map[topic]])

    def search(self, query: str, limit: int = 5) -> List[Dict[str, Any]]:
        """Enkel fritextsökning."""
        results = []
        query_lower = query.lower()
        for chunk in self.chunks:
            if query_lower in chunk['text'].lower() or query_lower in chunk['header_path'].lower():
                results.append({
                    "topic": chunk['header_path'],
                    "snippet": chunk['text'][:300] + "..." if len(chunk['text']) > 300 else chunk['text'],
                })
        return results[:limit]

File: src/library.ts
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

// Get current directory equivalent to __dirname in ESM
const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

export interface Chunk {
  id: number;
  text: string;
  header_path: string;
  metadata: {
    page: number | null;
    origin: string;
  };
}

export interface SearchResult {
  topic: string;
  snippet: string;
}

export class ADKLibrary {
  private chunks: Chunk[] = [];
  private headerMap: Map<string, Chunk[]> = new Map();
  private dataPath: string;

  constructor(dataPath?: string) {
    if (dataPath) {
      this.dataPath = dataPath;
    } else {
      // Default path relative to this file in dist/ or src/
      // If we are in dist/index.js, data is in ../src/data or similar?
      // Let's assume the data file is copied to dist/data or remains in src/data
      // Ideally, we copy it to dist/data during build.
      this.dataPath = path.join(__dirname, 'data', 'adk_chunks.jsonl');
    }
    this.loadData();
  }

  private loadData() {
    try {
      if (!fs.existsSync(this.dataPath)) {
        console.error(`File not found: ${this.dataPath}`);
        return;
      }

      const fileContent = fs.readFileSync(this.dataPath, 'utf-8');
      const lines = fileContent.split('\n');

      for (const line of lines) {
        if (!line.trim()) continue;
        try {
          const chunk: Chunk = JSON.parse(line);
          this.chunks.push(chunk);
          
          const headerPath = chunk.header_path || 'Unknown';
          if (!this.headerMap.has(headerPath)) {
            this.headerMap.set(headerPath, []);
          }
          this.headerMap.get(headerPath)?.push(chunk);
        } catch (e) {
          console.warn('Could not parse line in JSONL file');
        }
      }
      console.error(`Loaded ${this.chunks.length} chunks covering ${this.headerMap.size} topics.`);
    } catch (error) {
      console.error(`Error loading data from ${this.dataPath}:`, error);
    }
  }

  public getStructure(): string[] {
    return Array.from(this.headerMap.keys()).sort();
  }

  public readTopic(topic: string): string {
    if (!this.headerMap.has(topic)) {
      return `Error: Topic '${topic}' not found.`;
    }
    
    const chunks = this.headerMap.get(topic);
    if (!chunks) return "";
    
    return chunks.map(c => c.text).join('\n\n');
  }

  public search(query: string, limit: number = 5): SearchResult[] {
    const queryLower = query.toLowerCase();
    const results: SearchResult[] = [];

    for (const chunk of this.chunks) {
      if (
        chunk.text.toLowerCase().includes(queryLower) || 
        chunk.header_path.toLowerCase().includes(queryLower)
      ) {
        results.push({
          topic: chunk.header_path,
          snippet: chunk.text.length > 300 ? chunk.text.substring(0, 300) + '...' : chunk.text
        });
      }
    }

    return results.slice(0, limit);
  }
}

